src/visualizers/FactorFence.ts

```diff
@@ -0,0 +1,738 @@
+    private scaleFactor = 1.0 // zooming
+    private first = 0 // first term
+    private last = 0 // last term
+    private heightScale = 55 // stretching
+
+    // for vertical scaling, store max/min sequence vals displayed
+    private maxVal = 1
+    private minVal = 1
+
+    // lower left corner of graph
+    private graphCorner = new p5.Vector()
+    // lower left corner of text
+    private textCorner = new p5.Vector()
+
+    // vector shift from one bar to the next
+    private recSpace = new p5.Vector()
+    private recWidth = 12 // horizontal shift
+
+    // text control
+    private textInterval = 0
+    private textSize = 0
+
+    // for issues of caching taking time
+    private collectFailed = false
+
+    // colour palette
+    private palette = new factorPalette()
+
+    // frame counter for timeout on loop()
+    // first factorization failure
+    private frame = 0
+    private firstFailure = Number.POSITIVE_INFINITY
+
+    barsShowing() {
+        // determine which terms will be on the screen
+        // in order to decide how to initialize the graph
+        // and to be efficient in computing only the portion shown
+
+        // minimum bar to compute
+        // no less than first term
+        const minBars = Math.max(
+            Math.floor(
+                this.first - Math.min(0, this.graphCorner.x / this.recSpace.x)
+            ) - 1,
+            this.first
+        )
+
+        // number of bars on screen
+        // two extra to slightly bleed over edge
+        const numBars = Math.floor(
+            Math.min(
+                this.sketch.width / this.scaleFactor / this.recSpace.x + 2,
+                this.last - minBars
+            )
+        )
+
+        // maximum bar to compute
+        const maxBars = minBars + numBars
+
+        return {
+            minBars: minBars,
+            maxBars: maxBars,
+            numBars: numBars,
+        }
+    }
+
+    checkParameters(params: ParamValues<typeof paramDesc>) {
+        const status = super.checkParameters(params)
+
+        if (this.seq.last < this.terms) {
+            this.terms = this.seq.last
+        }
+
+        return status
+    }
+
+    collectDataForScale(barsInfo: barsData) {
+        // collect some info on the sequence in order to decide
+        // how best to vertically scale the initial graph
+        this.collectFailed = false
+        const seqVals = Array.from(Array(barsInfo.numBars), (_, i) => {
+            let elt = 1n
+            try {
+                elt = this.seq.getElement(i + barsInfo.minBars)
+            } catch {
+                this.collectFailed = true
+            }
+            let sig = 1n // sign of element
+            if (elt < 0n && this.signs) {
+                sig = -1n
+            }
+            // can I simplify this??
+            if (elt * sig < 1n) {
+                // in case elt = 0, store 1
+                elt = sig
+            }
+            // store height of bar (log) but with sign info
```

Comment:
I am confused by this code. Suppose this.signs is false and the sequence is entirely negative. Then doesn't this return all bar heights equal to 1, and doesn't that throw off your scaling?

Reply:
Such a sequence is A002819.  You're correct and I've refactored.

Reply:
committed